src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -91,23 +91,26 @@ public FormPipeReader(PipeReader pipeReader, Encoding encoding)
 
                 var buffer = readResult.Buffer;
 
-                if (!buffer.IsEmpty)
-                {
-                    ParseFormValues(ref buffer, ref accumulator, readResult.IsCompleted);
-                }
-
-                if (readResult.IsCompleted)
+                try
                 {
-                    _pipeReader.AdvanceTo(buffer.End);
-
                     if (!buffer.IsEmpty)
                     {
-                        throw new InvalidOperationException("End of body before form was fully parsed.");
+                        ParseFormValues(ref buffer, ref accumulator, readResult.IsCompleted);
                     }
-                    break;
-                }
 
-                _pipeReader.AdvanceTo(buffer.Start, buffer.End);
+                    if (readResult.IsCompleted)
+                    {
+                        if (!buffer.IsEmpty)
+                        {
+                            throw new InvalidOperationException("End of body before form was fully parsed.");
+                        }
+                        break;
+                    }
+                }
+                finally
```

Comment:
This should address #17723

Reply:
Note: rebase on the 3.1 fix and reconcile. https://github.com/dotnet/aspnetcore/pull/18939/files#diff-f44d0d010602be96c8821603bcd529ccR95. The refactor here is a minor cleanup.

Reply:
Don't forget about this!